diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java
index 886fde6..cf9ce84 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java
@@ -30,10 +30,10 @@ import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.classification.InterfaceAudience;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.hbase.MasterNotRunningException;
-import org.apache.hadoop.hbase.ZNodeClearer;
 import org.apache.hadoop.hbase.HConstants;
 import org.apache.hadoop.hbase.LocalHBaseCluster;
+import org.apache.hadoop.hbase.MasterNotRunningException;
+import org.apache.hadoop.hbase.ZNodeClearer;
 import org.apache.hadoop.hbase.ZooKeeperConnectionException;
 import org.apache.hadoop.hbase.client.HBaseAdmin;
 import org.apache.hadoop.hbase.regionserver.HRegionServer;
@@ -52,6 +52,7 @@ public class HMasterCommandLine extends ServerCommandLine {
     " start  Start Master. If local mode, start Master and RegionServer in same JVM\n" +
     " stop   Start cluster shutdown; Master signals RegionServer shutdown\n" +
     " clear  Delete the master znode in ZooKeeper after a master crashes\n "+
+    " autorestart  clear and start. Useful for supervisor-like scripts\n" +
     " where [opts] are:\n" +
     "   --minRegionServers=<servers>   Minimum RegionServers needed to host user tables.\n" +
     "   --localRegionServers=<servers> " +
@@ -136,6 +137,9 @@ public class HMasterCommandLine extends ServerCommandLine {
       return stopMaster();
     } else if ("clear".equals(command)) {
       return (ZNodeClearer.clear(getConf()) ? 0 : 1);
+    } else if ("autorestart".equals(command)) {
+      ZNodeClearer.clear(getConf());
+      return startMaster();
     } else {
       usage("Invalid command: " + command);
       return 1;
